openlibrary/plugins/openlibrary/js/SearchBar.js

```diff
@@ -0,0 +1,298 @@
+                </a>
+            </li>`;
+    }
+}
+
+/**
+ * Manages the interactions associated with the search bar in the header
+ */
+export class SearchBar {
+    /**
+     * @param {HTMLElement|JQuery}
+     * @param {Object?} urlParams
+     */
+    constructor(component, urlParams={}) {
+        /** UI Elements */
+        this.$component = $(component);
+        this.$form = this.$component.find('form.search-bar-input');
+        this.$input = this.$form.find('input[type="text"]');
+        this.$results = this.$component.find('ul.search-results');
+        this.$facetSelect = this.$component.find('.search-facet-selector select');
+
+        /** State */
+        /** Whether the bar is in collapsible mode */
+        this.inCollapsibleMode = false;
+        /** Whether the search bar is currently collapsed */
+        this.collapsed = false;
+        /** Selected facet (persisted) */
+        this.facet = new PersistentValue('facet', {
+            default: DEFAULT_FACET,
+            initValidation(val) { return val in FACET_TO_ENDPOINT; }
+        });
+
+        this.initFromUrlParams(urlParams);
+        this.initCollapsibleMode();
+
+        // Bind to changes in the search state
+        SearchUtils.mode.change(this.handleSearchModeChange.bind(this));
+        this.facet.change(this.handleFacetValueChange.bind(this));
+        this.$facetSelect.change(this.handleFacetSelectChange.bind(this));
+        this.$form.on('submit', this.submitForm.bind(this));
+
+        this.initAutocompletionLogic();
+    }
+
+    /** @type {String} The endpoint of the active facet */
+    get facetEndpoint() {
+        return FACET_TO_ENDPOINT[this.facet.read()];
+    }
+
+    /**
+     * Update internal state from url parameters
+     * @param {Object} urlParams
+     */
+    initFromUrlParams(urlParams) {
+        if (urlParams.facet in FACET_TO_ENDPOINT) {
+            this.facet.write(urlParams.facet);
+        }
+
+        if (urlParams.q) {
+            let q = urlParams.q.replace(/\+/g, ' ');
+            if (this.facet.read() === 'title' && q.indexOf('title:') != -1) {
+                const parts = q.split('"');
+                if (parts.length === 3) {
+                    q = parts[1];
+                }
+            }
+            this.$input.val(q);
+        }
+    }
+
+    submitForm() {
+        if (this.facet.read() === 'title') {
+            const q = this.$input.val();
+            this.$input.val(SearchBar.marshalBookSearchQuery(q));
+        }
+        this.$form.attr('action', this.composeSearchUrl(this.$input.val()));
+        SearchUtils.addModeInputsToForm(this.$form, SearchUtils.mode.read());
+    }
+
+    /** Initialize event handlers that allow the form to collapse for small screens */
+    initCollapsibleMode() {
+        this.toggleCollapsibleModeForSmallScreens($(window).width());
+        $(window).resize(debounce(() => {
```

Comment:
Since this is outside the scope of the SearchBar I think this should be bound outside the class. Consider reusing the subscriber pattern you have used elsewhere.

```
cont sb = new SearchBar();
$(window).resize(debounce(() => {
  sb.emit( 'resize' );
} );

Reply:
I think this is pretty important and if we are going to spend time refactoring components I think we need to setup some guidelines of what we are trying to achieve. 

This is a big anti-pattern that caused us a lot of problems in MediaWiki's mobile site and I'd rather not see that mistake repeated. I've started collecting these thoughts here: https://github.com/internetarchive/openlibrary/wiki/Component-Library#components-must-be-scoped-and-cannot-have-side-effects-on-things-outside-their-scope

Reply:
Same as https://github.com/internetarchive/openlibrary/pull/2285#discussion_r453084742 ; I'm not crazy over this proposed solution because it requires creating/exposing options which will never be modified.